Plugins cannot be loaded on debug builds

If the workspace that contains the `pluginlib` and `class_loader` packages are compiled with `catkin_make -DCMAKE_BUILD_TYPE="Debug"` command, the plugins cannot be loaded.

For example trying to load `image_transport/raw_pub` plugin produces the following logs

```
[DEBUG] [1379221062.565758000]: Attempting to create managed instance for class image_transport/raw_pub.
[DEBUG] [1379221062.565798000]: Class image_transport/raw_pub maps to library lib/libimage_transport_plugins in classes_available_.
[DEBUG] [1379221062.671369000]: Iterating through all possible paths where lib/libimage_transport_plugins could be located...
[DEBUG] [1379221062.673023000]: Checking path /Users/murat/ros/common_ws/devel/lib/lib/libimage_transport_pluginsd.dylib 
[DEBUG] [1379221062.673180000]: Checking path /Users/murat/ros/common_ws/devel/lib//libimage_transport_pluginsd.dylib 
[DEBUG] [1379221062.673327000]: Checking path /Users/murat/ros/core_ws/devel/lib/lib/libimage_transport_pluginsd.dylib 
[DEBUG] [1379221062.673495000]: Checking path /Users/murat/ros/core_ws/devel/lib//libimage_transport_pluginsd.dylib 
[DEBUG] [1379221062.674935000]: No path could be found to the library containing image_transport/raw_pub.
[ WARN] [1379221062.675168000]: Failed to load plugin image_transport/raw_pub, error string: Could not find library corresponding to plugin image_transport/raw_pub. Make sure the plugin description XML file has the correct name of the library and that the library actually exists.
```

I figured this is because the debug builds are linked against debug version of Poco library, which uses `d.dylib` suffix for shared libraries on `OS X`. On the other hand all plugin libraries have `.dylib` suffix even if they are compiled using `Debug` build type.

`Poco 1.4.6p1` is installed on the system via `Homebrew`

I occasionally need to build ros packages using `Debug` build type to debug other problems. This issue comes up whenever I do that.

I just ran into this again. On my machine, when `class_loader` and `pluginlib` are built with build type Debug, my `rviz` cannot locate plugins:

```
[DEBUG] [1398842515.000109000]: Attempting to create UNMANAGED instance for class rviz/Orbit.
[DEBUG] [1398842515.000203000]: Class rviz/Orbit maps to library libdefault_plugin in classes_available_.
[DEBUG] [1398842515.058778000]: Iterating through all possible paths where libdefault_plugin could be located...
[DEBUG] [1398842515.058922000]: Checking path /tmp/rviz_ws/devel/lib/libdefault_plugind.dylib
[DEBUG] [1398842515.059005000]: Checking path /tmp/rviz_ws/devel/lib/libdefault_plugind.dylib
[DEBUG] [1398842515.059058000]: Checking path /tmp/rviz_ws/underlay/install/lib/libdefault_plugind.dylib
[DEBUG] [1398842515.059116000]: Checking path /tmp/rviz_ws/underlay/install/lib/libdefault_plugind.dylib
[DEBUG] [1398842515.059182000]: Checking path /tmp/rviz_ws/src/rviz/libdefault_plugind.dylib
[DEBUG] [1398842515.059239000]: Checking path /tmp/rviz_ws/src/rviz/libdefault_plugind.dylib
[DEBUG] [1398842515.059293000]: No path could be found to the library containing rviz/Orbit.
[ERROR] [1398842515.059378000]: PluginlibFactory: The plugin for class 'rviz/Orbit' failed to load.  Error: Could not find library corresponding to plugin rviz/Orbit. Make sure the plugin description XML file has the correct name of the library and that the library actually exists.
```

As you can see it is looking for `/tmp/rviz_ws/src/rviz/libdefault_plugind.dylib` which has an extra `d` before the `.dylib`. The file `/tmp/rviz_ws/src/rviz/libdefault_plugin.dylib` exists on my machine.

This appears to be because of how `class_loader` uses `Poco`:

https://github.com/ros/class_loader/blob/hydro-devel/src/class_loader.cpp#L44

According to `Poco` (http://www.appinf.com/docs/poco/Poco.SharedLibrary.html):

> suffix static
> static std::string suffix();
> Returns the platform-specific filename suffix for shared libraries (including the period).
> In debug mode, the suffix also includes a "d" to specify the debug version of a library.

I think that `pluginlib` should be made more intelligent here:

https://github.com/ros/pluginlib/blob/indigo-devel/include/pluginlib/class_loader_imp.h#L238-261

Such that it checks for the extra `d` and if it finds in the suffix returned by `class_loader` then it should try paths both with and without the `d`.

@esteve this is a pretty important bug to fix, can you take a look at it soonish?

--- a/include/pluginlib/class_loader_imp.h
+++ b/include/pluginlib/class_loader_imp.h
@@ -247,8 +247,17 @@ namespace pluginlib
     std::vector<std::string> all_paths;
     std::vector<std::string> all_paths_without_extension = getCatkinLibraryPaths();
     all_paths_without_extension.push_back(getROSBuildLibraryPath(exporting_package_name));
-    std::string library_name_with_extension = library_name + class_loader::systemLibrarySuffix();
-    std::string stripped_library_name_with_extension = stripAllButFileFromPath(library_name) + class_loader::systemLibrarySuffix();
+    bool debug_library_suffix = (class_loader::systemLibrarySuffix().compare(0, 1, "d") == 0);
+    std::string non_debug_suffix;
+    if(debug_library_suffix) {
+        non_debug_suffix = class_loader::systemLibrarySuffix().substr(1);
+    } else {
+        non_debug_suffix = class_loader::systemLibrarySuffix();
+    }
+    std::string library_name_with_extension = library_name + non_debug_suffix;
+    std::string stripped_library_name = stripAllButFileFromPath(library_name);
+    std::string stripped_library_name_with_extension = stripped_library_name + non_debug_suffix;
+
     const std::string path_separator = getPathSeparator();
 
     for(unsigned int c = 0; c < all_paths_without_extension.size(); c++)
@@ -256,7 +265,13 @@ namespace pluginlib
       std::string current_path = all_paths_without_extension.at(c);
       all_paths.push_back(current_path + path_separator + library_name_with_extension);
       all_paths.push_back(current_path + path_separator + stripped_library_name_with_extension);
+      // We're in debug mode, try debug libraries as well
+      if(debug_library_suffix) {
+          all_paths.push_back(current_path + path_separator + library_name + class_loader::systemLibrarySuffix());
+          all_paths.push_back(current_path + path_separator + stripped_library_name + class_loader::systemLibrarySuffix());
+      }
     }
+
     return(all_paths);
   }
 

